age/src/ssh.rs

```diff
@@ -152,123 +152,6 @@ mod decrypt {
-            len_bytes
-                .iter()
-                .fold(Ok(0usize), |acc, x| {
-                    acc.and_then(|acc| {
-                        acc.checked_shl(8)
-                            .ok_or_else(|| {
-                                nom::Err::Failure(make_error(mid, ErrorKind::LengthValue))
-                            })
-                            .map(|acc| acc + (*x as usize))
-                    })
-                })
-                .map(|result| (i, result))
-        }
-    }
-
-    fn integer(input: &[u8]) -> IResult<&[u8], BigUint> {
-        preceded(
-            // Type: Universal | Primitive | INTEGER
-            der_type(0, 0, 2),
-            map(length_data(der_length), BigUint::from_bytes_be),
-        )(input)
-    }
-
-    fn tag_version(ver: u8) -> impl Fn(&[u8]) -> IResult<&[u8], &[u8]> {
-        move |input: &[u8]| {
-            preceded(
-                // Type: Universal | Primitive | INTEGER
-                der_type(0, 0, 2),
-                length_value(
-                    map_opt(der_length, |ver_len| match ver_len {
-                        1 => Some(ver_len),
-                        _ => None,
-                    }),
-                    tag(&[ver]),
-                ),
-            )(input)
-        }
-    }
-
-    /// A PKCS#1-encoded RSA private key.
-    ///
-    /// From [RFC 8017](https://tools.ietf.org/html/rfc8017#appendix-A.1.2):
-    /// ```text
-    /// RSAPrivateKey ::= SEQUENCE {
-    ///  version           Version,
-    ///  modulus           INTEGER,  -- n
-    ///  publicExponent    INTEGER,  -- e
-    ///  privateExponent   INTEGER,  -- d
-    ///  prime1            INTEGER,  -- p
-    ///  prime2            INTEGER,  -- q
-    ///  exponent1         INTEGER,  -- d mod (p-1)
-    ///  exponent2         INTEGER,  -- d mod (q-1)
-    ///  coefficient       INTEGER,  -- (inverse of q) mod p
-    ///  otherPrimeInfos   OtherPrimeInfos OPTIONAL
-    /// }
-    /// ```
-    ///
-    /// We only support the two-prime encoding, where `version = 0` and `otherPrimeInfos`
-    /// is omitted.
-    pub(super) fn rsa_privkey(input: &[u8]) -> IResult<&[u8], rsa::RsaPrivateKey> {
-        preceded(
-            // Type: Universal | Constructed | SEQUENCE
-            der_type(0, 1, 16),
-            length_value(
-                der_length,
-                preceded(
-                    tag_version(0),
-                    terminated(
-                        map(
-                            tuple((integer, integer, integer, integer, integer)),
-                            |(n, e, d, p, q)| {
-                                rsa::RsaPrivateKey::from_components(n, e, d, vec![p, q])
-                            },
-                        ),
-                        // d mod (p-1), d mod (q-1), iqmp
-                        tuple((integer, integer, integer)),
-                    ),
-                ),
-            ),
-        )(input)
```

Comment:
This fix has been merged, and released in `pkcs1 0.3`. However, that version also bumped the MSRV to 1.56, so we cannot use it here until we similarly want to raise our MSRV. This also affects the `rsa` crate, which does not have a release depending on `pkcs1 0.3` yet: https://github.com/RustCrypto/RSA/pull/123

Reply:
I've bumped the MSRV to 1.56, so we can use `pkcs1 0.3` now. There still isn't a new `rsa` release yet though.

Reply:
There’s a prerelease if that’s helpful: https://crates.io/crates/rsa/0.6.0-pre

Reply:
That's caused me to discover a different problem: I cannot upgrade to `rsa 0.6.0-pre` because it requires `num-bigint 0.8.1`, which in turn requires `zeroize 1.5`, but `chacha20poly1305 0.9` requires `zeroize >=1 <1.5`.

Reply:
There’s a `chacha20poly1305` prerelease as well: https://crates.io/crates/chacha20poly1305/0.10.0-pre

Reply:
We now depend on `rsa 0.7`, which depends on `pkcs1 0.4`, which should include the necessary fix.